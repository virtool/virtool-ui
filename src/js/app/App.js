import { ConnectedRouter } from "connected-react-router";
import React, { Suspense } from "react";
import { connect, Provider } from "react-redux";
import { ThemeProvider } from "styled-components";
import { WallContainer } from "../wall/Container";
import Reset from "../wall/Reset";
import { GlobalStyles } from "./GlobalStyles";
import Main from "./Main";
import { theme } from "./theme";

const LazyFirstUser = React.lazy(() => import("../wall/FirstUser"));
const LazyLogin = React.lazy(() => import("../wall/Login"));

export const App = ({ first, login, reset }) => {
    if (first) {
        return (
            <Suspense fallback={<WallContainer />}>
                <LazyFirstUser />
            </Suspense>
        );
    }

    if (login) {
        return (
            <Suspense fallback={<WallContainer />}>
                <LazyLogin />
            </Suspense>
        );
    }

    if (reset) {
        return <Reset />;
    }

    return <Main />;
};

export const mapStateToProps = state => {
    const { first, login, reset } = state.app;
    return {
        first,
        login,
        reset
    };
};

const ConnectedApp = connect(mapStateToProps)(App);

export default ({ store, history }) => {
    return (
        <ThemeProvider theme={theme}>
            <Provider store={store}>
                <ConnectedRouter history={history}>
                    <GlobalStyles />
                    <ConnectedApp />
                </ConnectedRouter>
            </Provider>
        </ThemeProvider>
    );
};
