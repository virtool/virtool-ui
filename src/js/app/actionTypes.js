/**
 * Create a special action type used for requests.
 *
 * The request is an object with three properties assigned with action types used for API requests.
 *
 * - REQUESTED
 * - SUCCEEDED
 * - FAILED
 *
 * @param root {string} the root name of the action type
 * @returns {object} a request-style action type
 */
const createRequestActionType = root => ({
    REQUESTED: `${root}_REQUESTED`,
    SUCCEEDED: `${root}_SUCCEEDED`,
    FAILED: `${root}_FAILED`,
});

// App
export const GET_INITIAL_STATE = createRequestActionType("GET_INITIAL_STATE");

// Account
export const GET_ACCOUNT = createRequestActionType("GET_ACCOUNT");
export const UPDATE_ACCOUNT = createRequestActionType("UPDATE_ACCOUNT");
export const CHANGE_ACCOUNT_PASSWORD = createRequestActionType("CHANGE_ACCOUNT_PASSWORD");
export const LOGIN = createRequestActionType("LOGIN");
export const LOGOUT = createRequestActionType("LOGOUT");
export const RESET_PASSWORD = createRequestActionType("RESET_PASSWORD");

// Administration Settings
export const GET_SETTINGS = createRequestActionType("GET_SETTINGS");
export const UPDATE_SETTINGS = createRequestActionType("UPDATE_SETTINGS");

// Analysis
export const WS_UPDATE_ANALYSIS = "WS_UPDATE_ANALYSIS";
export const SET_ANALYSIS = "SET_ANALYSIS";
export const SET_ANALYSIS_ACTIVE_ID = "SET_ANALYSIS_ACTIVE_ID";
export const SET_SEARCH_IDS = "SET_SEARCH_IDS";
export const SET_AODP_FILTER = "ADD_AODP_FILTER";
export const SET_ANALYSIS_SORT_KEY = "SET_ANALYSIS_SORT_KEY";

// Errors
export const CLEAR_ERROR = "CLEAR_ERROR";

// Files
export const UPLOAD = createRequestActionType("UPLOAD");
export const UPLOAD_SAMPLE_FILE = createRequestActionType("UPLOAD_SAMPLE_FILE");
export const UPLOAD_PROGRESS = "UPLOAD_PROGRESS";
export const UPLOAD_FAILED = "UPLOAD_FAILED";
export const REMOVE_UPLOAD = "REMOVE_UPLOAD";

// Forms
export const SET_PERSISTENT_FORM_STATE = "SET_PERSISTENT_FORM_STATE";
export const DELETE_PERSISTENT_FORM_STATE = "DELETE_PERSISTENT_FORM_STATE";
export const SET_REDUX_FORM_STATE = "SET_REDUX_FORM_STATE";

// Groups
export const CREATE_GROUP = createRequestActionType("CREATE_GROUP");

// HMMs
export const GET_HMM = createRequestActionType("GET_HMM");

// Indexes
export const GET_INDEX = createRequestActionType("GET_INDEX");
export const CREATE_INDEX = createRequestActionType("CREATE_INDEX");

// Jobs
export const GET_JOB = createRequestActionType("GET_JOB");

// OTU
export const GET_OTU = createRequestActionType("GET_OTU");
export const CREATE_OTU = createRequestActionType("CREATE_OTU");
export const EDIT_OTU = createRequestActionType("EDIT_OTU");
export const ADD_ISOLATE = createRequestActionType("ADD_ISOLATE");
export const EDIT_ISOLATE = createRequestActionType("EDIT_ISOLATE");
export const ADD_SEQUENCE = createRequestActionType("ADD_SEQUENCE");
export const EDIT_SEQUENCE = createRequestActionType("EDIT_SEQUENCE");

// Refs
export const GET_REFERENCE = createRequestActionType("GET_REFERENCE");

// Samples
export const GET_SAMPLE = createRequestActionType("GET_SAMPLE");
export const CREATE_SAMPLE = createRequestActionType("CREATE_SAMPLE");
export const UPDATE_SAMPLE = createRequestActionType("UPDATE_SAMPLE");

// Subtraction
export const GET_SUBTRACTION = createRequestActionType("GET_SUBTRACTION");
export const CREATE_SUBTRACTION = createRequestActionType("CREATE_SUBTRACTION");

// Users
export const FIND_USERS = createRequestActionType("FIND_USERS");
export const GET_USER = createRequestActionType("GET_USER");
export const CREATE_USER = createRequestActionType("CREATE_USER");
export const CREATE_FIRST_USER = createRequestActionType("CREATE_FIRST_USER");
export const EDIT_USER = createRequestActionType("EDIT_USER");
