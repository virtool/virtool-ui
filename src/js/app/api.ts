import { Request } from "./request";

export const root = () => Request.get("/");
