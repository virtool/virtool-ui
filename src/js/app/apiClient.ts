import Superagent, { Request } from "superagent";

const agent = Superagent.agent();

function prefixRequestUrl(request: Request) {
    if (request.url[0] !== "/") {
        request.url = `/${request.url}`;
    }

    if (!request.url.startsWith("/api")) {
        request.url = `/api${request.url}`;
    }

    return request;
}

agent.accept("application/json");
agent.use(prefixRequestUrl);

export const apiClient = agent;
