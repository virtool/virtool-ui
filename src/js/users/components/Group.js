import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";
import { Checkbox, SelectBoxGroupSection } from "../../base";

const StyledUserGroup = styled(SelectBoxGroupSection)`
    text-transform: capitalize;
    user-select: none;
`;

export class UserGroup extends React.Component {
    static propTypes = {
        id: PropTypes.string.isRequired,
        toggled: PropTypes.bool.isRequired,
        onClick: PropTypes.func.isRequired
    };

    handleClick = () => {
        this.props.onClick(this.props.id);
    };

    render() {
        return (
            <StyledUserGroup active={this.props.toggled} onClick={this.handleClick}>
                <Checkbox checked={this.props.toggled} label={this.props.id} />
            </StyledUserGroup>
        );
    }
}
