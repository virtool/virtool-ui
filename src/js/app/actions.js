import { PUSH_STATE, SET_INITIAL_STATE } from "./actionTypes";

export const pushState = state => ({
    type: PUSH_STATE,
    state
});

export const setInitialState = ({ dev, first_user }) => ({
    type: SET_INITIAL_STATE,
    dev,
    first: first_user
});
