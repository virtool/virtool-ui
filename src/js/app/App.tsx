import { LoadingPlaceholder } from "@/base";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { WallContainer } from "@wall/components/Container";
import { useAuthentication, useRootQuery } from "@wall/queries";
import { History } from "history";
import React, { Suspense } from "react";
import { Router } from "react-router-dom";
import { CompatRouter } from "react-router-dom-v5-compat";
import { ThemeProvider } from "styled-components";
import { GlobalStyles } from "./GlobalStyles";
import Main from "./Main";
import { theme } from "./theme";

// Lazy load components
const LazyFirstUser = React.lazy(() => import("@wall/components/FirstUser"));
const LazyLoginWall = React.lazy(() => import("@wall/components/LoginWall"));

/** The main application component that handles authentication and routing */
function ConnectedApp(): React.ReactElement {
    const { data: rootData, isPending: isRootPending } = useRootQuery();
    const { authenticated, isPending: isAuthPending } = useAuthentication();

    if (isRootPending || isAuthPending) {
        return <LoadingPlaceholder />;
    }

    if (rootData.first_user) {
        return (
            <Suspense fallback={<WallContainer />}>
                <LazyFirstUser />
            </Suspense>
        );
    }

    if (!authenticated) {
        return (
            <Suspense fallback={<WallContainer />}>
                <LazyLoginWall />
            </Suspense>
        );
    }

    return <Main />;
}

// Query client setup with default options and error handling
const queryClient = new QueryClient({
    defaultOptions: {
        queries: {
            retry: (failureCount: number, error: any) => {
                if ([401, 403, 404].includes(error.response?.status)) {
                    return false;
                }
                return failureCount <= 3;
            },
            staleTime: 2000,
        },
    },
});

type AppProps = {
    // React Router history object
    history: History;
};

/** The root App component that provides theme, query client, and routing setup */
export default function App({ history }: AppProps): React.ReactElement {
    return (
        <ThemeProvider theme={theme}>
            <QueryClientProvider client={queryClient}>
                <Router history={history}>
                    <CompatRouter>
                        <GlobalStyles />
                        <ConnectedApp />
                    </CompatRouter>
                </Router>
            </QueryClientProvider>
        </ThemeProvider>
    );
}
