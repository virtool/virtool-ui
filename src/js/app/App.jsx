import { ConnectedRouter } from "connected-react-router";
import React, { Suspense, useEffect } from "react";
import { QueryClient, QueryClientProvider } from "react-query";
import { connect, Provider } from "react-redux";
import { ThemeProvider } from "styled-components";
import { LoadingPlaceholder } from "../base";
import { resetClient } from "../utils/utils";
import { WallContainer } from "../wall/Container";
import Reset from "../wall/Reset";
import { getInitialState } from "./actions";
import { GlobalStyles } from "./GlobalStyles";
import Main from "./Main";
import { theme } from "./theme";

const LazyFirstUser = React.lazy(() => import("../wall/FirstUser"));
const LazyLogin = React.lazy(() => import("../wall/Login"));

function mapStateToProps(state) {
    const { first, login, reset, ready } = state.app;
    return {
        first,
        login,
        reset,
        ready,
    };
}

function mapDispatchToProps(dispatch) {
    return {
        getInitialState: () => {
            dispatch(getInitialState());
        },
    };
}

const ConnectedApp = connect(
    mapStateToProps,
    mapDispatchToProps,
)(({ first, login, reset, ready, getInitialState }) => {
    useEffect(getInitialState, []);
    if (!ready) {
        return <LoadingPlaceholder />;
    }

    if (first) {
        return (
            <Suspense fallback={<WallContainer />}>
                <LazyFirstUser />
            </Suspense>
        );
    }

    if (login) {
        return (
            <Suspense fallback={<WallContainer />}>
                <LazyLogin />
            </Suspense>
        );
    }

    if (reset) {
        return <Reset />;
    }

    return <Main />;
});

const queryClient = new QueryClient({
    defaultOptions: {
        queries: {
            onError: error => {
                if (error.response.status === 401) {
                    resetClient();
                }
            },
            retry: (failureCount, error) => {
                if (error.response.status === 401 || error.response.status === 404) {
                    return false;
                }
                return failureCount <= 3;
            },
        },
    },
});

export default function App({ store, history }) {
    return (
        <ThemeProvider theme={theme}>
            <QueryClientProvider client={queryClient}>
                <Provider store={store}>
                    <ConnectedRouter history={history}>
                        <GlobalStyles />
                        <ConnectedApp />
                    </ConnectedRouter>
                </Provider>
            </QueryClientProvider>
        </ThemeProvider>
    );
}
