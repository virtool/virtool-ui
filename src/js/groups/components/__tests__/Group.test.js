import { CHANGE_ACTIVE_GROUP } from "../../../app/actionTypes";
import { Group, mapDispatchToProps, mapStateToProps } from "../Group";

describe("Group", () => {
    let props;

    beforeEach(() => {
        props = {
            id: "foo",
            active: false,
            onSelect: jest.fn()
        };
    });

    it("should render correct with default props", () => {
        const wrapper = shallow(<Group {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should render as active when [active=true]", () => {
        props.active = true;
        const wrapper = shallow(<Group {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should call onSelect when clicked", () => {
        props.active = true;
        const wrapper = shallow(<Group {...props} />);
        wrapper.simulate("click");
        expect(props.onSelect).toHaveBeenCalled();
    });
});

describe("mapStateToProps", () => {
    const state = {
        groups: { activeId: "bar" }
    };

    it("should return [active=true] when active", () => {
        const result = mapStateToProps(state, { id: "bar" });
        expect(result).toEqual({
            active: true
        });
    });

    it("should return [active=false] when not active", () => {
        const result = mapStateToProps(state, { id: "foo" });
        expect(result).toEqual({
            active: false
        });
    });
});

describe("mapDispatchToProps", () => {
    it("should call dispatch with correct action", () => {
        const dispatch = jest.fn();
        const props = mapDispatchToProps(dispatch, { id: "foo" });

        props.onSelect();

        expect(dispatch).toHaveBeenCalledWith({
            type: CHANGE_ACTIVE_GROUP,
            id: "foo"
        });
    });
});
