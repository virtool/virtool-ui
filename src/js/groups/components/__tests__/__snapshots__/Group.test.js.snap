// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Group should render as active when [active=true] 1`] = `
<Group__StyledGroup
  active={true}
  key="foo"
  onClick={[MockFunction]}
>
  foo
</Group__StyledGroup>
`;

exports[`Group should render correct with default props 1`] = `
<Group__StyledGroup
  active={false}
  key="foo"
  onClick={[MockFunction]}
>
  foo
</Group__StyledGroup>
`;
